### ISSUE TITLE
Extract Foldable trait from Traversable

### ISSUE THREAD

**Issue by Badmoonz** [zio/zio-prelude#472]
Some kind of data structures can have constraint on type paramters, like Ord[A] for ordered structures like Set / TreeSet, but they are not Funktors (Covariant)
[](https://wiki.haskell.org/Foldable_and_Traversable#Foldable)

And Foldable by itself provides as lot of useful functionality even ```foreach_[G[+_]: IdentityBoth: Covariant, A](fa: F[A])(f: A => G[Any]): G[Unit]```

suggested trait: 
```
trait Foldable[A] {
   def foldLeft[S, A](fa: F[A])(s: S)(f: (S, A) => S): S   

  /// all other methods implemented using foldLeft
}
```

**Comment by sideeffffect** [comment]
What laws apply for Foldable?

**Comment by Badmoonz** [comment]
probably there are no fundamental laws 
we just can  check that overriden methods comute to default implementations
but we cannot provide lawful check that foldLeft iterrates over the structure in the right way

**Comment by Badmoonz** [comment]
withing Traversable we also can check that Foldable.foldLeft and implementation via foreach are equivalent

**Comment by sideeffffect** [comment]
That would make me thing that `Foldable` is not as much an actual Type class, but more like a "shape" (we talked about Shapes [here](https://github.com/zio/zio-prelude/pull/351))

**Comment by sideeffffect** [comment]
Although there are some laws in the [Haskell definition](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Foldable.html)

> ```haskell
> foldr f z t = appEndo (foldMap (Endo . f) t ) z
> 
> foldl f z t = appEndo (getDual (foldMap (Dual . Endo . flip f) t)) z
>
> fold = foldMap id
>
> length = getSum . foldMap (Sum . const  1)
> ```

Although [here](https://en.wikibooks.org/wiki/Haskell/Foldable) it says that

> `Foldable` is slightly unusual among Haskell classes which are both principled and general-purpose in that it has no laws of its own. The closest thing is the following property, which strictly speaking is not a law (as it is guaranteed to hold whatever the instance is): given a [monoid homomorphism](https://en.wikibooks.org/wiki/Haskell/Monoids#Homomorphisms) `g`,
>
> ```haskell
> foldMap (g . f) = g . foldMap f
> ```

**Comment by Badmoonz** [comment]
i will insvestigate the "shape" thread

and about the laws , as I said, it's just checks that overriden methods are equivalent to default implementations

**Comment by adamgfraser** [comment]
Yes, when we were putting the library together we did not include `Foldable` because it did not have meaningful laws but agree it would be good to think about how to abstract over data types like this.

**Comment by Badmoonz** [comment]
> Yes, when we were putting the library together we did not include `Foldable` because it did not have meaningful laws but agree it would be good to think about how to abstract over data types like this.

I think that Foldable  can have the same laws as Hash: folding Equal objects give Equal results

**Comment by jdegoes** [comment]
I think I'd rather handle this with a "subcategory" Traverse, which could include arbitrary constraints on the types.

**Comment by Badmoonz** [comment]
> I think I'd rather handle this with a "subcategory" Traverse, which could include arbitrary constraints on the types.

Set's are still not covariant even in subcategory; map's can change structure by squashing produced equal values
